components/storage/blockdevice/COMPONENT_OSPIF/OSPIFBlockDevice.cpp

```diff
@@ -0,0 +1,1908 @@
+    if (sfdp_detect_addressability(param_table, _sfdp_info.bptbl) < 0) {
+        tr_error("Verify 4byte addressing failed");
+        return -1;
+    }
+
+    if (sfdp_detect_device_density(param_table, _sfdp_info.bptbl) < 0) {
+        tr_error("Detecting device density failed");
+        return -1;
+    }
+
+    // Set Page Size (OSPI write must be done on Page limits)
+    _page_size_bytes = sfdp_detect_page_size(param_table, sfdp_info.bptbl.size);
+
+    if (_sfdp_detect_reset_protocol_and_reset(param_table) != OSPIF_BD_ERROR_OK) {
+        tr_error("Init - Detecting reset protocol/resetting failed");
+        return -1;
+    }
+
+    // Detect and Set Erase Types
+    bool shouldSetQuadEnable = false;
+    bool is_qpi_mode = false;
+    bool is_opi_mode = false;
+
+    if (sfdp_detect_erase_types_inst_and_size(param_table, _sfdp_info) < 0) {
+        tr_error("Init - Detecting erase types instructions/sizes failed");
+        return -1;
+    }
+
+    // Detect and Set fastest Bus mode (default 1-1-1)
+    _sfdp_detect_best_bus_read_mode(param_table, sfdp_info.bptbl.size, shouldSetQuadEnable, is_qpi_mode, is_opi_mode);
+    if (true == is_opi_mode) {
+        change_mode(DOPI);
+        tr_debug("Init - Setting DOPI mode");
+    }
+    if (true == shouldSetQuadEnable) {
+        if (_needs_fast_mode) {
+            _enable_fast_mode();
+        }
+        // Set Quad Enable and QPI Bus modes if Supported
+        tr_debug("Init - Setting Quad Enable");
+        if (0 != _sfdp_set_quad_enabled(param_table)) {
+            tr_error("Device supports Quad bus, but Quad Enable Failed");
+            return -1;
+        }
+        if (true == is_qpi_mode) {
+            tr_debug("Init - Setting QPI mode");
+            _sfdp_set_qpi_enabled(param_table);
+        }
+    }
+
+#ifndef TARGET_NORDIC
+    // 4 byte addressing is not currently supported with the Nordic OSPI controller
+    if (_attempt_4_byte_addressing) {
+        if (_sfdp_detect_and_enable_4byte_addressing(param_table, sfdp_info.bptbl.size) != OSPIF_BD_ERROR_OK) {
+            tr_error("Init - Detecting/enabling 4-byte addressing failed");
+            return -1;
+        }
+    }
+#endif
+
+    if (false == _is_mem_ready()) {
+        tr_error("Init - _is_mem_ready Failed");
+        return -1;
+    }
+
+    return 0;
+}
+
+int OSPIFBlockDevice::_sfdp_set_quad_enabled(uint8_t *basic_param_table_ptr)
+{
+    uint8_t status_reg_setup[OSPI_MAX_STATUS_REGISTERS] = {0};
+    uint8_t status_regs[OSPI_MAX_STATUS_REGISTERS] = {0};
+
+    // QUAD Enable procedure is specified by 3 bits
+    uint8_t qer_value = (basic_param_table_ptr[OSPIF_BASIC_PARAM_TABLE_QER_BYTE] & 0x70) >> 4;
+
+    switch (qer_value) {
+        case 0:
+            tr_debug("Device Does not Have a QE Bit, continue based on Read Inst");
+            return 0;
+        case 1:
+        case 4:
+            // Bit 1 of Status Reg 2
+            _quad_enable_register_idx = 1;
```

Comment:
quad->octal, unless your planning to combine implementations between QSPI and OSPI

Reply:
we are planning to combine implementations between QSPI and OSPI.